### PR TITLE
fix(dlx): failures on concurrent runs of dlx

### PR DESCRIPTION
**Pull request by zkochan**
Closes https://github.com/pnpm/pnpm/pull/8420

### CODE REVIEW

File: exec/plugin-commands-script-runners/src/dlx.ts [hunk 1]
```diff
@@ -190,17 +193,28 @@ export function createCacheKey (pkgs: string[], registries: Record<string, strin
...
 }
 
-function isCacheValid (cacheLink: string, dlxCacheMaxAge: number): boolean {
+function getValidCacheDir (cacheLink: string, dlxCacheMaxAge: number): string | undefined {
   let stats: Stats
+  let target: string
   try {
     stats = fs.lstatSync(cacheLink)
+    if (stats.isSymbolicLink()) {
+      target = fs.readlinkSync(cacheLink)
```
Comment (KSXGitHub) [inline exec/plugin-commands-script-runners/src/dlx.ts]:
```suggestion
      target = fs.realpathSync(cacheLink)
```

File: exec/plugin-commands-script-runners/src/dlx.ts [hunk 2]
```diff
@@ -190,17 +193,28 @@ export function createCacheKey (pkgs: string[], registries: Record<string, strin
...
     if (util.types.isNativeError(err) && 'code' in err && err.code === 'ENOENT') {
-      return false
+      return undefined
     }
     throw err
   }
-  return stats.mtime.getTime() + dlxCacheMaxAge * 60_000 >= new Date().getTime()
+  const isValid = stats.mtime.getTime() + dlxCacheMaxAge * 60_000 >= new Date().getTime()
+  if (isValid) {
+    return path.resolve(path.dirname(cacheLink), target)
```
Comment (KSXGitHub) [inline exec/plugin-commands-script-runners/src/dlx.ts]: Assuming you applied the above suggestion,

```suggestion
    return target
```

**Review by KSXGitHub — APPROVED** [review]
I do have some suggestions.

File: exec/plugin-commands-script-runners/src/dlx.ts [hunk 3]
```diff
@@ -190,17 +193,25 @@ export function createCacheKey (pkgs: string[], registries: Record<string, strin
...
 
-function isCacheValid (cacheLink: string, dlxCacheMaxAge: number): boolean {
+function getValidCacheDir (cacheLink: string, dlxCacheMaxAge: number): string | undefined {
   let stats: Stats
+  let target: string
   try {
     stats = fs.lstatSync(cacheLink)
+    if (stats.isSymbolicLink()) {
+      target = fs.realpathSync(cacheLink)
+      if (!target) return undefined
```
Comment (KSXGitHub) [inline exec/plugin-commands-script-runners/src/dlx.ts:203]: Is this line necessary?

Comment (zkochan) [reply exec/plugin-commands-script-runners/src/dlx.ts:203]: no, but it is weird to allow the validity check below if the target doesn't exist.

**Pull request merged**